src/completions/History.ts

```diff
@@ -75,14 +95,54 @@ export class HistoryCompletionSource extends Completions.CompletionSourceFuse {
         }
         options += options ? " " : ""
 
+        const searchURLs = config.get("searchurls")
+
+        this.options = []
+        this.updateSectionHeader("History and bookmarks")
+        const tokens = query.split(" ")
+        if (tokens.length > 1 || query.endsWith(" ")) {
+            // Not on first token
+            const su = tokens[0]
+            if (Object.prototype.hasOwnProperty.call(searchURLs, su)) {
+                query = tokens.slice(1).join(" ")
+                const url = UrlUtil.interpolateSearchItem(
+                    new URL(searchURLs[su]),
+                    query,
+                )
+                this.updateSectionHeader("Search " + su + " (" + url.href + ")")
+
+                // Actual query sent to browser needs to be space separated
+                // list of tokens, otherwise partial matches won't be found
+                query = searchURLs[su].split("%s").join(" ") + " " + query
+            }
+        } else {
+            // Still on first token, so display matching search urls
+            for (const prop in searchURLs) {
+                if (Object.prototype.hasOwnProperty.call(searchURLs, prop)) {
+                    if (this.options.length == 5) {
+                        break
+                    }
+                    if (prop.startsWith(query)) {
+                        this.options.push(
+                            new SearchUrlCompletionOption(
+                                prop,
+                                searchURLs[prop],
+                            ),
+                        )
+                    }
+                }
+            }
+        }
+
         // Options are pre-trimmed to the right length.
         // Typescript throws an error here - further investigation is probably warranted
-        this.options = ((await this.scoreOptions(
-            query,
-            config.get("historyresults"),
-        )) as any).map(
-            page => new HistoryCompletionOption(options + page.url, page),
-        )
+        const histOptions = (
+            (await this.scoreOptions(
+                query,
+                config.get("historyresults"),
+            )) as any
+        ).map(page => new HistoryCompletionOption(options + page.url, page))
+        this.options = this.options.concat(histOptions)
```

Comment:
I've set the default to `searchurlweight=150` for now, so searchurls show up higher than bookmarks. Happy to change that if you think bookmarks should have priority.

Reply:
By the way, I noticed that the frecency function just returns the visit count at the moment. It looks like frecency would be pretty easy to implement using `browser.getHistory.getVisits()` - should I give that a go, or is there a reason that's been left?

Reply:
I think I never bothered to implement it because I found that frequency was fine on its own in practice.

I would happily merge a PR that implemented it and left it configurable :)